bundles/org.openhab.binding.mercedesme/src/main/java/org/openhab/binding/mercedesme/internal/server/CallbackServer.java

```diff
@@ -0,0 +1,190 @@
+    private static final Map<Integer, CallbackServer> SERVER_MAP = new HashMap<Integer, CallbackServer>();
+
+    private Optional<Server> server = Optional.empty();
+    private Optional<AccessTokenResponse> token = Optional.empty();
+    private AccessTokenRefreshListener listener;
+    private AccountConfiguration config;
+    private OAuthClientService oacs;
+    private String callbackUrl;
+
+    public CallbackServer(AccessTokenRefreshListener l, HttpClient hc, OAuthFactory oAuthFactory,
+            AccountConfiguration config, String callbackUrl) {
+        oacs = oAuthFactory.createOAuthClientService(Constants.OAUTH_CLIENT_NAME + config.callbackPort,
+                Constants.MB_TOKEN_URL, Constants.MB_AUTH_URL, config.clientId, config.clientSecret, config.getScope(),
+                false);
+        AUTH_MAP.put(Integer.valueOf(config.callbackPort), oacs);
+        SERVER_MAP.put(Integer.valueOf(config.callbackPort), this);
+        listener = l;
+        this.config = config;
+        this.callbackUrl = callbackUrl;
+    }
+
+    public String getAuthorizationUrl() {
+        try {
+            return oacs.getAuthorizationUrl(callbackUrl, null, null);
+        } catch (OAuthException e) {
+            LOGGER.error("Error creating Authorization URL {}", e.getMessage());
+            return Constants.EMPTY;
+        }
+    }
+
+    public String getScope() {
+        return config.getScope();
+    }
+
+    public void start() {
+        LOGGER.debug("Start Callback Server for port {}", config.callbackPort);
+        if (!server.isEmpty()) {
+            LOGGER.debug("Callback server for port {} already started", config.callbackPort);
+            return;
+        }
+        server = Optional.of(new Server());
+        ServerConnector connector = new ServerConnector(server.get());
+        connector.setPort(config.callbackPort);
+        server.get().setConnectors(new Connector[] { connector });
+        ServletHandler servletHandler = new ServletHandler();
+        server.get().setHandler(servletHandler);
+        servletHandler.addServletWithMapping(CallbackServlet.class, Constants.CALLBACK_ENDPOINT);
+        try {
+            server.get().start();
+        } catch (Exception e) {
+            LOGGER.warn("Cannot start Callback Server for port {}, Error {}", config.callbackPort, e.getMessage());
+        }
+    }
+
+    public void stop() {
+        LOGGER.debug("Stop Callback Server");
+        try {
+            if (!server.isEmpty()) {
+                server.get().stop();
+                server = Optional.empty();
+            }
+        } catch (Exception e) {
+            LOGGER.warn("Cannot start Callback Server for port {}, Error {}", config.callbackPort, e.getMessage());
+        }
+    }
+
+    public String getToken() {
+        if (token.isEmpty()) {
+            LOGGER.debug("Token empty - Manual Authorization needed at {}", callbackUrl);
```

Comment:
Either this is probably the wrong place for logging this, or the method name is misleading. I would expect a get() method to return something, not log anything or contain much logic.

Reply:
Changed in rework commit